truffle/src/main/java/org/jruby/truffle/core/format/format/FormatCharacterNode.java

```diff
@@ -0,0 +1,96 @@
+import org.jruby.truffle.core.format.convert.ToIntegerNodeGen;
+import org.jruby.truffle.core.format.convert.ToStringNode;
+import org.jruby.truffle.core.format.convert.ToStringNodeGen;
+import org.jruby.truffle.core.format.exceptions.NoImplicitConversionException;
+import org.jruby.truffle.core.format.write.bytes.WriteByteNodeGen;
+import org.jruby.truffle.language.control.RaiseException;
+
+import java.nio.charset.StandardCharsets;
+
+@NodeChildren({
+    @NodeChild(value = "width", type = FormatNode.class),
+    @NodeChild(value = "value", type = FormatNode.class),
+})
+public abstract class FormatCharacterNode extends FormatNode {
+
+    private final boolean hasMinusFlag;
+
+    @Child private ToIntegerNode toIntegerNode;
+    @Child private ToStringNode toStringNode;
+
+    public FormatCharacterNode(RubyContext context, boolean hasMinusFlag) {
+        super(context);
+        this.hasMinusFlag = hasMinusFlag;
+    }
+
+    @Specialization
+    protected byte[] format(VirtualFrame frame, int width, Object value) {
+        if (toStringNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            toStringNode = insert(ToStringNodeGen.create(getContext(),
+                false,
+                "to_str",
+                false,
+                null,
+                WriteByteNodeGen.create(getContext(), new LiteralFormatNode(getContext(), value))));
+        }
+        Object toStrResult;
+        try {
+            toStrResult = toStringNode.executeToString(frame, value);
+        } catch (NoImplicitConversionException e) {
+            toStrResult = null;
+        }
+
+        final String charString;
+        if (toStrResult == null || isNil(toStrResult)) {
+            if (toIntegerNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                toIntegerNode = insert(ToIntegerNodeGen.create(getContext(), null));
+            }
+            final int charValue = (int) toIntegerNode.executeToInteger(frame, value);
+            // TODO BJF check char length is > 0
+            charString = Character.toString((char) charValue);
+        } else {
+            final String resultString = new String((byte[]) toStrResult);
+            final int size = resultString.length();
+            if (size > 1) {
+                throw new RaiseException(getContext().getCoreExceptions().argumentErrorCharacterRequired(this));
+            }
+            charString = resultString;
+        }
+
+        final boolean leftJustified = hasMinusFlag || width < 0;
+        if (width < 0) {
+            width = -width;
+        }
+
+        return doFormat(charString, width, leftJustified).getBytes(StandardCharsets.US_ASCII);
+    }
+
+    @TruffleBoundary
+    private String doFormat(final String charString, final int width, final boolean leftJustified) {
+        return String.format("%" + (leftJustified ? "-" : "") + width + "." + width + "s", charString);
```

Comment:
Can you just cache this generated format string? Doing that is trivial and should be an easy performance win until we implement our own formatting code.